packages/filesystem/src/browser/file-tree/file-tree-model.ts

```diff
@@ -148,7 +149,13 @@ export class FileTreeModel extends TreeModelImpl implements LocationService {
     }
 
     async copy(source: URI, target: Readonly<FileStatNode>): Promise<URI> {
-        const targetUri = target.uri.resolve(source.path.base);
+        let targetUri = target.uri.resolve(source.path.base);
+        // duplicate file or folder
+        if (source.path.toString() === target.uri.path.toString()) {
+            const parent = await this.fileService.resolve(source.parent);
+            const name = source.path.name + '_copy';
+            targetUri = FileSystemUtils.generateUniqueResourceURI(source.parent, parent, name, source.path.ext);
+        }
         try {
             await this.fileService.copy(source, targetUri);
```

Comment:
I believe we should move the logic to the actual implementation of `fileService.copy`, so all clients (callers) who are requesting a similar behavior will not need to duplicate the logic. For example, the `doMoveCopy` is handling the `_copy` like you've done:

https://github.com/eclipse-theia/theia/blob/5db308fd83745536bdbeec162172b6c40f35c9e7/packages/filesystem/src/browser/file-service.ts#L1120-L1125

Reply:
> I believe we should move the logic to the actual implementation of `fileService.copy`

I do not think so. If you debug into the VS Code's `FileService`, that we copied, you will see the target URI is the expected target URI (`file:///path/to/file copy` or `file:///path/to/file copy 2`). Code does not do any more magic just copies/moves the file. We should do the same since we copied the code and we do not want to diverge from it.

By the way, I proposed something similar [here](https://github.com/eclipse-theia/theia/pull/9037#issuecomment-775905295)